ansible_base/rbac/api/serializers.py

```diff
@@ -0,0 +1,232 @@
+from django.conf import settings
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import transaction
+from django.db.utils import IntegrityError
+from django.utils.translation import gettext_lazy as _
+from rest_framework import serializers
+from rest_framework.exceptions import PermissionDenied
+from rest_framework.fields import flatten_choices_dict, to_choices_dict
+from rest_framework.serializers import ValidationError
+
+from ansible_base.lib.abstract_models.common import get_url_for_object
+from ansible_base.lib.serializers.common import CommonModelSerializer
+from ansible_base.lib.utils.validation import ansible_id_validator
+from ansible_base.rbac.models import RoleDefinition, RoleTeamAssignment, RoleUserAssignment
+from ansible_base.rbac.permission_registry import permission_registry  # careful for circular imports
+from ansible_base.rbac.validators import validate_permissions_for_model
+
+
+class ChoiceLikeMixin(serializers.ChoiceField):
+    """
+    This uses a ForeignKey to populate the choices of a choice field.
+    This also manages some string manipulation, right now, adding the local service name.
+    """
+
+    default_error_messages = serializers.PrimaryKeyRelatedField.default_error_messages
+    psuedo_field = None  # define in subclass
+
+    def get_model_for_init(self):
+        "Delay any model references until serializer initialization due to load order concerns"
+        raise NotImplementedError
+
+    def get_dynamic_choices(self):
+        raise NotImplementedError
+
+    def get_dynamic_object(self, data):
+        raise NotImplementedError
+
+    def to_representation(self, value):
+        raise NotImplementedError
+
+    def __init__(self, **kwargs):
+        # Workaround so that the parent class does not resolve the choices right away
+        self.html_cutoff = kwargs.pop('html_cutoff', self.html_cutoff)
+        self.html_cutoff_text = kwargs.pop('html_cutoff_text', self.html_cutoff_text)
+
+        self.allow_blank = kwargs.pop('allow_blank', False)
+        super(serializers.ChoiceField, self).__init__(**kwargs)
+
+    def _initialize_choices(self):
+        choices = self.get_dynamic_choices()
+        self._grouped_choices = to_choices_dict(choices)
+        self._choices = flatten_choices_dict(self._grouped_choices)
+        self.choice_strings_to_values = {str(k): k for k in self._choices}
+
+    @property
+    def grouped_choices(self):
+        if not hasattr(self, '_grouped_choices'):
+            self._initialize_choices()
+        return self._grouped_choices
+
+    @property
+    def choices(self):
```

Comment:
We should probably use https://docs.djangoproject.com/en/5.0/ref/utils/#django.utils.functional.cached_property for these.

Reply:
This didn't really accomplish anything, since you are still manually managing an attribute to be the cache.  The point about my comment was that `@cached_property` would remove the need to have `self._choices`.

Reply:
I pushed a change to clean that up a little bit, and further lean into reliance on `cached_property`. And I hope I can get it cleaned up a good deal more with the custom permission model later.